### DISCUSSION TITLE
v0.11 Dev Channel

### DISCUSSION THREAD

**Discussion by Blinue** [Blinue/Magpie#512] [Dev Channel]
~鉴于架构已经成熟，是时候发布 1.0 了。1.0 将聚焦于：~ 
架构依然不稳定，下个版本将是 0.11

~1. 支持打包。尝试发布到微软商店。~
2. 支持限制捕获帧率，这可以有效降低功耗。
3. 支持触控输入 #218

**Comment by Blinue** [comment]
经过漫长的等待，Weblate 的自由托管申请通过了

![image](https://user-images.githubusercontent.com/34770031/227908756-be589716-4799-4fc0-b1e9-99b2e92402b6.png)

**Comment by Blinue** [comment]
## 新的渲染架构

```mermaid
flowchart LR
    subgraph 呈现线程
    呈现((呈现))
    共享纹理{共享纹理}
    共享纹理--读取-->游戏内叠加层
    游戏内叠加层-->光标
    光标-->呈现
    呈现-->游戏内叠加层
    end
    subgraph 渲染线程
    捕获((捕获))
    捕获-->应用效果
    应用效果--复制到-->共享纹理
    应用效果-->捕获
    end
```

1. 两个 D3D 设备，两个渲染管线。一个负责捕获和应用效果，另一个负责绘制游戏内叠加层和光标以及呈现到屏幕。
2. 对于渲染线程
    * 应用效果可能非常昂贵
    * 渲染完成后复制到共享纹理
    * 可以限制捕获帧率以降低功耗，不会影响光标渲染
3. 对于呈现线程
    * 会以高于屏幕刷新率的速度循环以降低光标延迟
4. **渲染线程和呈现线程可以使用不同的显卡**。集成显卡负责呈现，独立显卡负责渲染。我相信这可以有效提高性能。
5. 对于轻量的效果不会提高性能，反而会消耗更多显存，因此不适合 3D 游戏。

**Comment by Blinue** [comment]
为了防止误操作，今后 Weblate 的更改会以 Pull request 的形式合并。

**Comment by hooke007** [comment]
> 尝试发布到微软商店。

如果要做这个还要加点导出功能，比如log（不然汇报bug找log能难倒一批人）

商店应用在受保护的隐藏目录，这个路径对普通人来说太阴间，而且一些应用也因为这个限制在商店的版本是功能残缺版。

**Reply by Blinue** [reply]
想必有很多限制，会一一解决的，有很多开源项目可以参考。

**Comment by eriforce** [comment]
关于3D Game Mode，因为在我的使用经验里我只知道它会影响鼠标能否移出当前屏幕边界以及是否可以和Overlay交互，是否开启它只和我是不是有多个屏幕以及我是不是需要去点其他屏幕或者和Overlay交互和有关，所以想请教一下是基于什么考虑把它放在Profile的配置里，会不会作为一个全局设置更合适？

**Reply by Blinue** [reply]
这个选项是为 3D 游戏设计的，会导致功能降级，非 3D 游戏不要开启。和窗口绑定是很自然的，如果窗口是 3D 游戏就打开它。

**Reply by eriforce** [reply]
`3D 游戏` 是不是指鼠标控制视角的第一人称游戏那种？

我个人使用中觉得不太方便的两个地方：
- 平时不希望鼠标移出当前屏幕，偶尔想去其他屏幕上点两下或者滚动一下网页
- 想鼠标不被overlay干扰、主要画面不被遮挡同时观察fps，需要先手动拖动一下overlay的时候

**Reply by Blinue** [reply]
可以移出屏幕比不能更好。每个人的习惯不同，你可以自己做更改。如果想观察 FPS，打开“显示帧率”即可。

**Reply by eriforce** [reply]
> 如果想观察 FPS，打开“显示帧率”即可。

嗯，这个我知道。我的意思是在不需要开3D模式的应用里，如果想要拖一下overlay会需要退出来开了再关。

目前我加了个快捷键这样全屏下也能直接开关。

**Comment by Blinue** [comment]
Magpie 已经可以通过 Scoop 安装了：https://github.com/ScoopInstaller/Extras/pull/10305

也可以通过 Chocolatey 安装：https://community.chocolatey.org/packages/magpie

**Comment by Blinue** [comment]
现在 Magpie 支持的语言越来越多了，应该更加国际化。README 默认使用英文，发行说明也会有英文版本。我主要用 ChatGPT 翻译成英语

**Comment by BLACKCHARGE** [comment]
现在完成了自定义任务栏后，左侧导航栏是怎么展开和收起的，快捷键吗，找不到以前点击收放的地方了😂

**Reply by Blinue** [reply]
不能收起了，Win11 的设置应用也是这么做的

**Comment by Blinue** [comment]
我正在考虑集成 [ONNX Runtime](https://github.com/microsoft/onnxruntime)，它有和 D3D11 交互的接口所以集成很方便。这个库现在似乎在兼容性和性能上有点问题，你有了解吗？@hooke007

**Reply by hooke007** [reply]
不算了解，我只过用它的dml（dx12）接口

**Reply by Blinue** [reply]
现在有什么能实时推理的模型？

**Reply by hooke007** [reply]
大多数复杂架构都没有实时的可能性，esrgan的简化架构算一个，看pth体积基本在4mb以下的，RTX4080可以做到实时（当然这个实时也仅仅是针对普通视频的帧率而言，以游戏动辄上百的标准来看几乎没有），

**Reply by Blinue** [reply]
基于 ML 肯定就不能对帧率有要求了，新架构实现之后即使帧率很低也算可以接受。以 C++ 集成意味着在推理以外没有性能损失，我们不需要在 CPU 和 GPU 间传递帧数据，因此效率会比 VapourSynth 更高。希望以后能有更轻量级的模型。

**Reply by hooke007** [reply]
刚测了一下如果是720x2的话还是可用的，基于目前我能找到的最快2x模型  (171.45 fps)

**Reply by Blinue** [reply]
有30帧我就能接受了。以后我会实验一下

**Comment by Blinue** [comment]
我正在测试新的实现方式，使源窗口成为缩放窗口的“所有者窗口”([Owner window](https://learn.microsoft.com/en-us/windows/win32/winmsg/window-features#owned-windows))，OS 将确保缩放窗口始终位于源窗口上方，这意味着缩放窗口不必置顶！还有一个好处是切换窗口后不必退出缩放，源窗口回到前台时 OS 将确保缩放窗口自动显示并置于源窗口上方，这对于窗口化缩放很有用。

**Reply by Blinue** [reply]
主要问题是和源窗口的耦合性更高了，更有可能导致兼容性问题。
1. 缩放 UWP 窗口时缩放窗口无法遮盖任务栏，这可能是因为 shell 特殊处理 UWP 窗口
2. 部分自定义了光标的窗口有时会有奇怪的光标闪烁

**Reply by Blinue** [reply]
前者可以使用 [TreatAsDesktopFullscreen](https://github.com/dechamps/WindowInvestigator/issues/3) 属性解决。

后者是因为源窗口的输入被附加到了缩放窗口上，这是所有者窗口的默认行为，见 https://devblogs.microsoft.com/oldnewthing/20130412-00/?p=4683 ，使用 AttachThreadInput 取消附加即可。

**Reply by Blinue** [reply]
这个方案和 Desktop Duplication 捕获不兼容。Desktop Duplication 要求源窗口必须在最前，因此也和窗口化不兼容，我认为是时候删除这个捕获方式了。

**Comment by Blinue** [comment]
#721 不见了，作者也注销了。标题是“希望根据窗口分辨率切换配置”，我在其中讨论了可能的实现方式，没了有点可惜。

**Reply by hooke007** [reply]
可尝试向gh客服申诉恢复意外删除的issue

**Reply by Blinue** [reply]
帖子本身没有很重要的东西，只是用来追踪进度。这倒提醒了我作者随时可以删除帖子无需所有者同意，所以不要把重要的东西放在别人的帖子里。（这个功能本身就不合理，帖子属于参与讨论的每一个人，作者私自删除会给社区造成损失）

**Reply by AcerolaShinobu** [reply]
当时是因为用户名问题被封号，后面又申述回来了

**Reply by Blinue** [reply]
敏感得过分了，哪里都逃不开政治啊
